File: 01_code.py
from collections import defaultdict

input_file = "01_input.txt"

first_list = []
second_list = []

with open(input_file, "r") as f:
    for line in f:
        first, second = line.split()
        first_list.append(int(first))
        second_list.append(int(second))

first_list.sort()
second_list.sort()

total_distance = 0
for first, second in zip(first_list, second_list):
    distance = abs(first - second)
    total_distance += distance

print(f"First part: {total_distance}")

count_dict = defaultdict(int)
for second in second_list:
    count_dict[second] += 1

total_score = 0
for first in first_list:
    total_score += count_dict[first] * first

print(f"Second part: {total_score}")

File: 01_code.cpp
#include <algorithm>
#include <iostream>
#include <fstream>
#include <sstream>
#include <vector>
#include <unordered_map>


int main() {
    const std::string filename = "01_input.txt";

    std::ifstream file(filename, std::ios::in);
    if (!file.is_open()) {
        std::cerr << "Failed to open the file." << std::endl;
        return 1;
    }

    std::size_t line_count = std::count(
        std::istreambuf_iterator<char>(file),
        std::istreambuf_iterator<char>(),
        '\n'
    );
    file.close();

    std::vector<int> first_list;
    std::vector<int> second_list;

    first_list.reserve(line_count);
    second_list.reserve(line_count);

    file.open(filename, std::ios::in);
    if (!file.is_open()) {
        std::cerr << "Failed to open the file." << std::endl;
        return 1;
    }

    std::string line;
    while (std::getline(file, line)) {
        std::istringstream stream(line); // Create a stream for the line

        int first_number;
        int second_number;

        stream >> first_number;
        stream >> second_number;
        first_list.push_back(first_number);
        second_list.push_back(second_number);
    }
    file.close();

    std::ranges::sort(first_list);
    std::ranges::sort(second_list);

    std::unordered_map<int, int> second_list_map;

    int total_distance = 0;
    for (int i = 0; i < line_count; i++) {
        second_list_map[second_list[i]]++;
        int distance = std::abs(first_list[i] - second_list[i]);
        total_distance += distance;
    }

    int total_score = 0;
    for (int num : first_list) {
        total_score += num * second_list_map[num];
    }

    std::cout << "First part: " << total_distance << std::endl;
    std::cout << "Second part: " << total_score << std::endl;

    return 0;
}
